flink-staging/flink-ml/src/main/scala/org/apache/flink/ml/optimization/GradientDescent.scala

```diff
@@ -0,0 +1,237 @@
+  *                      [[IterativeSolver.Iterations]] for the maximum number of iteration,
+  *                      [[IterativeSolver.Stepsize]] for the learning rate used.
+  */
+class GradientDescent(runParameters: ParameterMap) extends IterativeSolver {
+
+  import Solver.WEIGHTVECTOR_BROADCAST
+
+  var parameterMap: ParameterMap = parameters ++ runParameters
+
+  // TODO(tvas): Use once we have proper sampling in place
+//  case object MiniBatchFraction extends Parameter[Double] {
+//    val defaultValue = Some(1.0)
+//  }
+//
+//  def setMiniBatchFraction(fraction: Double): GradientDescent = {
+//    parameterMap.add(MiniBatchFraction, fraction)
+//    this
+//  }
+
+  /** Performs one iteration of Stochastic Gradient Descent using mini batches
+    *
+    * @param data A Dataset of LabeledVector (label, features) pairs
+    * @param currentWeights A Dataset with the current weights to be optimized as its only element
+    * @return A Dataset containing the weights after one stochastic gradient descent step
+    */
+  private def SGDStep(data: DataSet[(LabeledVector)], currentWeights: DataSet[WeightVector]):
+    DataSet[WeightVector] = {
+
+    // TODO: Sample from input to realize proper SGD
+    data.map {
+      new GradientCalculation
+    }.withBroadcastSet(currentWeights, WEIGHTVECTOR_BROADCAST).reduce {
+      (left, right) =>
+        val (leftGradientVector, leftCount) = left
+        val (rightGradientVector, rightCount) = right
+
+        BLAS.axpy(1.0, leftGradientVector.weights, rightGradientVector.weights)
+        (new WeightVector(
+          rightGradientVector.weights,
+          leftGradientVector.intercept + rightGradientVector.intercept),
+          leftCount + rightCount)
+    }.map {
+      new WeightsUpdate
+    }.withBroadcastSet(currentWeights, WEIGHTVECTOR_BROADCAST)
+  }
+
+  /** Provides a solution for the given optimization problem
+    *
+    * @param data A Dataset of LabeledVector (label, features) pairs
+    * @param initWeights The initial weights that will be optimized
+    * @return The weights, optimized for the provided data.
+    */
+  override def optimize(data: DataSet[LabeledVector], initWeights: Option[DataSet[WeightVector]]):
+  DataSet[WeightVector] = {
+    // TODO: Faster way to do this?
+    val dimensionsDS = data.map(_.vector.size).reduce((a, b) => b)
+
+    val numberOfIterations: Int = parameterMap(Iterations)
+
+
+    val initialWeightsDS: DataSet[WeightVector] = initWeights match {
+      //TODO(tvas): Need to figure out if and where we want to pass initial weights
+      case Some(x) => x
+      case None => createInitialWeightVector(dimensionsDS)
+    }
+
+    // We need the weights vector to initialize the regularization value.
+    val initWeightsVector = initialWeightsDS.collect()(0).weights
+
+    val initRegVal = parameterMap(RegularizationType)
+      .applyRegularization(initWeightsVector, 0.0, parameterMap(RegularizationParameter))._2
```

Comment:
For what do we need the `initRegVal` value? Why do we need the `initRegVal` on the client and not in a `DataSet`?

Introducing a `collect` call for this seems not to be worthwhile at the moment.